File: Tools/AiArtAgentPlatform/backend/app/schemas/style_pack.py
from __future__ import annotations

from typing import Literal

from pydantic import Field, field_validator, model_validator

from .core import (
    SLUG_PATTERN,
    AssetCategory,
    AssetTask,
    ImageOutputSpec,
    StrictModel,
)


class ReferenceSource(StrictModel):
    path: str = Field(min_length=1)
    mode: Literal["read_only"] = "read_only"

    @field_validator("path")
    @classmethod
    def require_absolute_path(cls, value: str) -> str:
        from pathlib import Path

        if not Path(value).is_absolute():
            raise ValueError("reference source path must be absolute")
        return value


class CameraStyle(StrictModel):
    projection: str = Field(min_length=1)
    pitch_semantic_min: int = Field(ge=0, le=90)
    pitch_semantic_max: int = Field(ge=0, le=90)
    shared_view_required: bool = True
    default_facing: str = Field(min_length=1)

    @model_validator(mode="after")
    def validate_pitch_range(self) -> CameraStyle:
        if self.pitch_semantic_min > self.pitch_semantic_max:
            raise ValueError("minimum camera pitch cannot exceed maximum camera pitch")
        return self


class PaletteStyle(StrictModel):
    base: list[str] = Field(min_length=1)
    accents: list[str] = Field(default_factory=list)


class RenderingStyle(StrictModel):
    character_proportion: str = Field(min_length=1)
    character_outline: str = Field(min_length=1)
    environment_detail: str = Field(min_length=1)
    surface_finish: str = Field(min_length=1)
    shadow_direction: str = Field(min_length=1)


class ReadabilityStyle(StrictModel):
    protect_playfield: bool = True
    character_contrast_above_environment: bool = True
    preserve_clear_silhouette: bool = True
    avoid_high_frequency_ground_noise: bool = True


class UiStyle(StrictModel):
    formal_text_baked_in: bool = False
    border_language: list[str] = Field(default_factory=list)


class StyleGuide(StrictModel):
    schema_version: Literal[1] = 1
    style_id: str = Field(min_length=1, max_length=160)
    display_name: str = Field(min_length=1, max_length=160)
    reference_source: ReferenceSource
    camera: CameraStyle
    palette: PaletteStyle
    rendering: RenderingStyle
    readability: ReadabilityStyle
    ui: UiStyle
    forbidden: list[str] = Field(default_factory=list)


class SourceReferenceFile(StrictModel):
    relative_path: str
    size_bytes: int = Field(ge=0)


class ReferenceImportRequest(StrictModel):
    reference_id: str = Field(pattern=SLUG_PATTERN)
    source_relative_path: str = Field(min_length=1, max_length=1000)
    categories: list[AssetCategory] = Field(min_length=1)
    identities: list[str] = Field(default_factory=list)
    usages: list[str] = Field(default_factory=list)
    viewpoints: list[str] = Field(default_factory=list)
    materials: list[str] = Field(default_factory=list)
    notes: str = Field(default="", max_length=2000)


class ReferenceUpdateRequest(StrictModel):
    categories: list[AssetCategory] = Field(min_length=1)
    identities: list[str] = Field(default_factory=list)
    usages: list[str] = Field(default_factory=list)
    viewpoints: list[str] = Field(default_factory=list)
    materials: list[str] = Field(default_factory=list)
    notes: str = Field(default="", max_length=2000)


class ReferenceAsset(StrictModel):
    reference_id: str = Field(pattern=SLUG_PATTERN)
    source_relative_path: str
    workspace_relative_path: str
    thumbnail_relative_path: str
    sha256: str = Field(pattern=r"^[a-f0-9]{64}$")
    width: int = Field(gt=0)
    height: int = Field(gt=0)
    categories: list[AssetCategory] = Field(min_length=1)
    identities: list[str] = Field(default_factory=list)
    usages: list[str] = Field(default_factory=list)
    viewpoints: list[str] = Field(default_factory=list)
    materials: list[str] = Field(default_factory=list)
    notes: str = ""


class ReferenceIndex(StrictModel):
    schema_version: Literal[1] = 1
    references: list[ReferenceAsset] = Field(default_factory=list)


class ReferenceFilters(StrictModel):
    category: AssetCategory | None = None
    identity: str | None = None
    usage: str | None = None
    viewpoint: str | None = None
    material: str | None = None
    limit: int = Field(default=100, ge=1, le=500)


class CharacterIdentity(StrictModel):
    asset_id: str = Field(pattern=SLUG_PATTERN)
    display_name: str = Field(min_length=1, max_length=120)
    silhouette: list[str] = Field(default_factory=list)
    face: list[str] = Field(default_factory=list)
    hair: list[str] = Field(default_factory=list)
    costume: list[str] = Field(default_factory=list)
    palette: list[str] = Field(default_factory=list)
    equipment: list[str] = Field(default_factory=list)
    immutable_traits: list[str] = Field(default_factory=list)


PromptSectionKey = Literal[
    "project_style",
    "asset_task",
    "identity",
    "references",
    "composition_camera",
    "lighting_materials",
    "output_spec",
    "forbidden",
    "postprocess",
]


class PromptSection(StrictModel):
    key: PromptSectionKey
    label: str
    content: str


class PromptPreviewRequest(StrictModel):
    task: AssetTask
    identity: CharacterIdentity | None = None
    viewpoint: str = ""
    composition: str = ""
    lighting: str = ""
    materials: list[str] = Field(default_factory=list)
    output_spec: ImageOutputSpec
    additional_negative_constraints: list[str] = Field(default_factory=list)
    prompt_override: str | None = None


class CompiledPrompt(StrictModel):
    task: AssetTask
    selected_reference_ids: list[str] = Field(max_length=4)
    sections: list[PromptSection] = Field(min_length=9, max_length=9)
    prompt: str = Field(min_length=1)
    negative_constraints: list[str]

File: Tools/AiArtAgentPlatform/frontend/src/api/stylePack.ts
import {
  useMutation,
  useQuery,
  useQueryClient,
} from "@tanstack/react-query";

import { deleteRequest, getJson, postJson, putJson } from "./client";
import type { AssetCategory } from "../types/core";

export interface StyleGuide {
  schema_version: 1;
  style_id: string;
  display_name: string;
  reference_source: {
    path: string;
    mode: "read_only";
  };
  camera: {
    projection: string;
    pitch_semantic_min: number;
    pitch_semantic_max: number;
    shared_view_required: boolean;
    default_facing: string;
  };
  palette: {
    base: string[];
    accents: string[];
  };
  rendering: {
    character_proportion: string;
    character_outline: string;
    environment_detail: string;
    surface_finish: string;
    shadow_direction: string;
  };
  readability: {
    protect_playfield: boolean;
    character_contrast_above_environment: boolean;
    preserve_clear_silhouette: boolean;
    avoid_high_frequency_ground_noise: boolean;
  };
  ui: {
    formal_text_baked_in: boolean;
    border_language: string[];
  };
  forbidden: string[];
}

export interface ReferenceAsset {
  reference_id: string;
  source_relative_path: string;
  workspace_relative_path: string;
  thumbnail_relative_path: string;
  sha256: string;
  width: number;
  height: number;
  categories: AssetCategory[];
  identities: string[];
  usages: string[];
  viewpoints: string[];
  materials: string[];
  notes: string;
}

export interface ReferenceImportInput {
  reference_id: string;
  source_relative_path: string;
  categories: AssetCategory[];
  identities: string[];
  usages: string[];
  viewpoints: string[];
  materials: string[];
  notes: string;
}

export type ReferenceUpdateInput = Pick<
  ReferenceAsset,
  | "categories"
  | "identities"
  | "usages"
  | "viewpoints"
  | "materials"
  | "notes"
>;

export interface SourceReferenceFile {
  relative_path: string;
  size_bytes: number;
}

export interface ReferenceFilters {
  category?: AssetCategory;
  identity?: string;
  usage?: string;
  viewpoint?: string;
  material?: string;
  limit?: number;
}

export interface AssetTaskInput {
  asset_id: string;
  category: AssetCategory;
  name: string;
  brief: string;
  usage: string;
  style_pack: string;
  reference_ids: string[];
  constraint_profile: string;
  constraint_overrides: Record<string, unknown>;
  candidate_count: number;
  output_mode: string;
}

export interface CharacterIdentityInput {
  asset_id: string;
  display_name: string;
  silhouette: string[];
  face: string[];
  hair: string[];
  costume: string[];
  palette: string[];
  equipment: string[];
  immutable_traits: string[];
}

export interface PromptPreviewInput {
  task: AssetTaskInput;
  identity: CharacterIdentityInput | null;
  viewpoint: string;
  composition: string;
  lighting: string;
  materials: string[];
  output_spec: {
    width: number;
    height: number;
    format: "png";
    transparent_required: boolean;
  };
  additional_negative_constraints: string[];
  prompt_override: string | null;
}

export interface PromptSection {
  key: string;
  label: string;
  content: string;
}

export interface CompiledPrompt {
  task: AssetTaskInput;
  selected_reference_ids: string[];
  sections: PromptSection[];
  prompt: string;
  negative_constraints: string[];
}

export function fetchStyleGuide(projectId: string): Promise<StyleGuide> {
  return getJson<StyleGuide>(`/api/v1/projects/${projectId}/style-guide`);
}

export function updateStyleGuide(
  projectId: string,
  guide: StyleGuide,
): Promise<StyleGuide> {
  return putJson<StyleGuide>(
    `/api/v1/projects/${projectId}/style-guide`,
    guide,
  );
}

export function fetchReferenceSource(
  projectId: string,
  query = "",
  limit = 100,
): Promise<SourceReferenceFile[]> {
  const params = new URLSearchParams({ query, limit: String(limit) });
  return getJson<SourceReferenceFile[]>(
    `/api/v1/projects/${projectId}/reference-source?${params.toString()}`,
  );
}

export function fetchReferences(
  projectId: string,
  filters: ReferenceFilters = {},
): Promise<ReferenceAsset[]> {
  const params = new URLSearchParams();
  if (filters.category) params.set("category", filters.category);
  if (filters.identity) params.set("identity", filters.identity);
  if (filters.usage) params.set("usage", filters.usage);
  if (filters.viewpoint) params.set("viewpoint", filters.viewpoint);
  if (filters.material) params.set("material", filters.material);
  if (filters.limit) params.set("limit", String(filters.limit));
  const suffix = params.size ? `?${params.toString()}` : "";
  return getJson<ReferenceAsset[]>(
    `/api/v1/projects/${projectId}/references${suffix}`,
  );
}

export function importReference(
  projectId: string,
  input: ReferenceImportInput,
): Promise<ReferenceAsset> {
  return postJson<ReferenceAsset>(
    `/api/v1/projects/${projectId}/references`,
    input,
  );
}

export function updateReference(
  projectId: string,
  referenceId: string,
  input: ReferenceUpdateInput,
): Promise<ReferenceAsset> {
  return putJson<ReferenceAsset>(
    `/api/v1/projects/${projectId}/references/${referenceId}`,
    input,
  );
}

export function deleteReference(
  projectId: string,
  referenceId: string,
): Promise<void> {
  return deleteRequest(
    `/api/v1/projects/${projectId}/references/${referenceId}`,
  );
}

export function referenceThumbnailUrl(
  projectId: string,
  reference: Pick<ReferenceAsset, "reference_id" | "sha256">,
): string {
  return (
    `/api/v1/projects/${projectId}/references/${reference.reference_id}` +
    `/thumbnail?v=${reference.sha256}`
  );
}

export function previewPrompt(
  projectId: string,
  input: PromptPreviewInput,
): Promise<CompiledPrompt> {
  return postJson<CompiledPrompt>(
    `/api/v1/projects/${projectId}/prompt-preview`,
    input,
  );
}

export function useStyleGuideQuery(projectId: string | undefined) {
  return useQuery({
    queryKey: ["style-guide", projectId],
    queryFn: () => fetchStyleGuide(projectId ?? ""),
    enabled: Boolean(projectId),
    retry: false,
    staleTime: 10_000,
  });
}

export function useUpdateStyleGuideMutation(projectId: string | undefined) {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (guide: StyleGuide) => updateStyleGuide(projectId ?? "", guide),
    onSuccess: async () => {
      await queryClient.invalidateQueries({ queryKey: ["style-guide", projectId] });
      await queryClient.invalidateQueries({
        queryKey: ["reference-source", projectId],
      });
    },
  });
}

export function useReferenceSourceQuery(
  projectId: string | undefined,
  query: string,
  limit: number,
) {
  return useQuery({
    queryKey: ["reference-source", projectId, query, limit],
    queryFn: () => fetchReferenceSource(projectId ?? "", query, limit),
    enabled: Boolean(projectId),
    retry: false,
  });
}

export function useReferencesQuery(
  projectId: string | undefined,
  filters: ReferenceFilters = {},
) {
  return useQuery({
    queryKey: ["references", projectId, filters],
    queryFn: () => fetchReferences(projectId ?? "", filters),
    enabled: Boolean(projectId),
    retry: false,
  });
}

export function useImportReferenceMutation(projectId: string | undefined) {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (input: ReferenceImportInput) =>
      importReference(projectId ?? "", input),
    onSuccess: async () => {
      await queryClient.invalidateQueries({ queryKey: ["references", projectId] });
      await queryClient.invalidateQueries({
        queryKey: ["project-activity", projectId],
      });
    },
  });
}

export function useUpdateReferenceMutation(projectId: string | undefined) {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: ({
      referenceId,
      input,
    }: {
      referenceId: string;
      input: ReferenceUpdateInput;
    }) => updateReference(projectId ?? "", referenceId, input),
    onSuccess: async () => {
      await queryClient.invalidateQueries({ queryKey: ["references", projectId] });
      await queryClient.invalidateQueries({
        queryKey: ["project-activity", projectId],
      });
    },
  });
}

export function useDeleteReferenceMutation(projectId: string | undefined) {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (referenceId: string) =>
      deleteReference(projectId ?? "", referenceId),
    onSuccess: async () => {
      await queryClient.invalidateQueries({ queryKey: ["references", projectId] });
      await queryClient.invalidateQueries({
        queryKey: ["project-activity", projectId],
      });
    },
  });
}

export function usePromptPreviewMutation(projectId: string | undefined) {
  return useMutation({
    mutationFn: (input: PromptPreviewInput) =>
      previewPrompt(projectId ?? "", input),
  });
}
